### PR TITLE
Add onSelectionEnd function parameter to CircularSlider

### PR DESCRIPTION
**Pull request by kirkeaton**
This pull request adds an additional callback function for when the slider selection has ended. This may be useful in cases where the value of the slider maybe be used to update a value on a server or database. In these cases one could use this function as opposed to onSelectionChange to avoid updating too often.

### CODE REVIEW

**Comment by davidanaya** [comment]
@kirkeaton thanks for your PR
Sorry I was not aware of it before and I did a few extra updates in the last days.
Could you please review it and fix the conflicts? Otherwise I will incorporate your suggestion myself when I have some time.

**Comment by kirkeaton** [comment]
@davidanaya 
I have just rebased this PR against the latest code base, fixed all of the conflicts, and added my changes to the SingleCircularSlider as well.

File: lib/src/circular_slider_paint.dart [hunk 1]
```diff
@@ -1,34 +1,54 @@
...
+  final CircularSliderMode mode;
   final int init;
   final int end;
-  final int intervals;
+  final int divisions;
   final int primarySectors;
   final int secondarySectors;
-  final Function onSelectionChange;
+  final SelectionChanged<int> onSelectionChange;
+  final Function onSelectionEnd;
```
Comment (davidanaya) [inline lib/src/circular_slider_paint.dart]: we could use the same typedef that I created for `onSelectionChange` (see line above)

Comment (davidanaya) [reply lib/src/circular_slider_paint.dart]: please note that we should also return the laps now

File: lib/src/circular_slider_paint.dart [hunk 2]
```diff
@@ -124,16 +238,46 @@ class _CircularSliderState extends State<CircularSliderPaint> {
...
+    
+    var angle = coordinatesToRadians(_painter.center, position);
+    var percentage = radiansToPercentage(angle);
+    var newValue = percentageToValue(percentage, widget.divisions);
+
+    if (_isInitHandlerSelected) {
+      widget.onSelectionEnd(newValue, widget.end);
+    } else {
+      widget.onSelectionEnd(widget.init, newValue);
+    }
```
Comment (davidanaya) [inline lib/src/circular_slider_paint.dart]: this is a bit more complex now
we could have both handlers selected (click on the selection and dragging moves the whole selection) so we would need to account for that (check lines 243 to 251 in this same file for reference)
also we would need to return also the laps (same lines)

File: lib/src/double_circular_slider.dart [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  /// color of the handlers
+  final Color handlerColor;
+
+  /// callback function when init and end change
+  /// (int init, int end) => void
+  final SelectionChanged<int> onSelectionChange;
+
+  /// callback function when init and end finish
+  /// (int init, int end) => void
+  final Function onSelectionEnd;
```
Comment (davidanaya) [inline lib/src/double_circular_slider.dart]: please use same type than with `onSelectionChange` above

File: lib/src/single_circular_slider.dart [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  /// color of the handlers
+  final Color handlerColor;
+
+  /// callback function when init and end change
+  /// (int init, int end) => void
+  final SelectionChanged<int> onSelectionChange;
+
+  /// callback function when init and end finish
+  /// (int init, int end) => void
+  final Function onSelectionEnd;
```
Comment (davidanaya) [inline lib/src/single_circular_slider.dart]: please use the same type as `onSelectionChange` above

File: lib/src/double_circular_slider.dart [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+          onSelectionChange: (newInit, newEnd, laps) {
+            if (widget.onSelectionChange != null) {
+              widget.onSelectionChange(newInit, newEnd, laps);
+            }
+            setState(() {
+              _init = newInit;
+              _end = newEnd;
+            });
+          },
+          onSelectionEnd: (newInit, newEnd) {
```
Comment (davidanaya) [inline lib/src/double_circular_slider.dart]: include laps

File: lib/src/single_circular_slider.dart [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+          onSelectionChange: (newInit, newEnd, laps) {
+            if (widget.onSelectionChange != null) {
+              widget.onSelectionChange(newInit, newEnd, laps);
+            }
+            setState(() {
+              _end = newEnd;
+            });
+          },
+          onSelectionEnd: (newInit, newEnd) {
+            if (widget.onSelectionEnd != null) {
```
Comment (davidanaya) [inline lib/src/single_circular_slider.dart:128]: include laps

**Review by davidanaya — CHANGES_REQUESTED** [review]
Thanks, it's looking good but there are a few things that need to be updated to incorporate the latest features in the library.

**Comment by kirkeaton** [comment]
@davidanaya Thanks for the review. I think I've addressed all the changes. I've consolidated the panning logic into a separate function to reduce duplicated code as well.

File: lib/src/circular_slider_paint.dart [hunk 7]
```diff
@@ -94,63 +138,181 @@ class _CircularSliderState extends State<CircularSliderPaint> {
...
+      return;
+    }
+
+    if (isSingleHandler) {
+      widget.onSelectionChange(widget.init, newValue, _laps);
+      return;
+    }
 
+    // isDoubleHandler but one handler was selected
     if (_isInitHandlerSelected) {
```
Comment (davidanaya) [inline lib/src/circular_slider_paint.dart:262]: I think this is unreachable unless you remove the previous code (lines 261 to 264)

**Review by davidanaya — CHANGES_REQUESTED** [review]
Only one thing missing, I think.

**Comment by kirkeaton** [comment]
@davidanaya Thanks, I went ahead and removed that if statement.

**Review by davidanaya — APPROVED** [review]
LGTM 🚀 
You can go ahead and merge it.
I'll test it tonight and if it works fine I'll do a release.
Thanks!

**Comment by kirkeaton** [comment]
@davidanaya Thank's for approving! I'm unable to merge as I don't have write access to the repository.

**Pull request merged**